File: agent/tools/builtin_tools/shell_tool.py
"""Shell 命令执行工具（白名单安全模式）"""
import re
import subprocess
from typing import Any
from tools.base_tool import BaseTool

# 允许执行的命令前缀白名单（只读/信息类）
_ALLOWED_PREFIXES = (
    # 基础信息
    "echo", "date", "time", "whoami", "hostname", "uname",
    # 文件/目录查看
    "ls", "dir", "pwd", "cat", "head", "tail", "wc", "find", "grep",
    # 网络
    "ping", "curl", "wget", "nslookup", "netstat", "ss",
    # 进程/系统信息
    "ps", "top", "df", "du", "free", "uptime", "env", "printenv",
    # 语言运行时
    "python", "python3", "node", "java -version",
    # Git 只读操作
    "git log", "git status", "git diff", "git branch", "git show", "git tag",
)

# 拒绝包含敏感文件路径的命令（防止泄露密钥、密码等）
_SENSITIVE_PATTERNS = re.compile(
    r"(\.env|id_rsa|id_ecdsa|\.pem|\.key|\.p12|shadow|passwd|credentials|secret)",
    re.IGNORECASE,
)


def _is_allowed(cmd: str) -> bool:
    c = cmd.strip().lower()
    return any(c.startswith(p) for p in _ALLOWED_PREFIXES)


def _has_sensitive_path(cmd: str) -> bool:
    """命令中含有敏感文件名关键词时拒绝执行。"""
    return bool(_SENSITIVE_PATTERNS.search(cmd))


class ShellTool(BaseTool):
    """安全 Shell 命令执行工具（白名单模式）。
    支持只读/信息类命令，禁止 rm/del/format 等破坏性操作。
    参数: command(要执行的命令,必填), timeout(超时秒数,默认10)
    """

    def __init__(self):
        super().__init__(
            description=(
                "执行安全的 Shell 命令并返回输出。"
                "参数: command(命令字符串,必填), timeout(超时秒数,默认10,最大30)"
            )
        )

    def execute(self, command: str, timeout: int = 10) -> dict[str, Any]:
        if not command or not command.strip():
            return {"success": False, "error": "command 不能为空"}

        if not _is_allowed(command):
            return {
                "success": False,
                "error": f"命令 '{command.split()[0]}' 不在允许列表中，仅支持只读/信息类命令",
            }

        if _has_sensitive_path(command):
            return {
                "success": False,
                "error": "命令包含敏感文件路径（.env / 密钥文件等），已拒绝执行",
            }

        timeout = min(int(timeout), 30)
        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=timeout,
                encoding="utf-8",
                errors="replace",
            )
            output = result.stdout or result.stderr or "(无输出)"
            if len(output) > 2000:
                output = output[:2000] + "\n...(输出已截断)"
            return {
                "success": result.returncode == 0,
                "returncode": result.returncode,
                "output": output,
            }
        except subprocess.TimeoutExpired:
            return {"success": False, "error": f"命令超时（>{timeout}s）"}
        except Exception as e:
            return {"success": False, "error": str(e)}

File: backend/web/src/main/java/com/intelligent/agent/web/ai/tool/builtin/shell/ShellTool.java
package com.intelligent.agent.web.ai.tool.builtin.shell;

import com.intelligent.agent.web.ai.tool.AgentTool;
import com.intelligent.agent.web.ai.tool.ToolDefinition;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

/**
 * 安全 Shell 工具（TODO-110 Task 1）：白名单只读/信息类命令，拒绝破坏性与敏感路径。
 */
public class ShellTool implements AgentTool {

    private static final List<String> ALLOWED_PREFIXES = List.of(
            "echo", "date", "time", "whoami", "hostname", "uname",
            "ls", "dir", "pwd", "cat", "head", "tail", "wc", "find", "grep",
            "ps", "df", "du", "free", "uptime", "env", "printenv",
            "git log", "git status", "git diff", "git branch", "git show", "git tag");
    private static final Pattern SENSITIVE = Pattern.compile(
            "\\.env|id_rsa|id_ecdsa|\\.pem|\\.key|\\.p12|shadow|passwd|credentials|secret",
            Pattern.CASE_INSENSITIVE);

    @Override
    public ToolDefinition definition() {
        return new ToolDefinition(
                "shell_tool", "执行安全的 Shell 命令（只读/信息类白名单）。"
                        + "参数: command(必填), timeout(秒,默认10,最大30)", true, null,
                Duration.ofSeconds(35),
                Map.of(
                        "type", "object",
                        "properties", Map.of(
                                "command", Map.of("type", "string", "description", "要执行的命令"),
                                "timeout", Map.of("type", "integer",
                                        "description", "超时秒数，默认 10，最大 30")),
                        "required", List.of("command")));
    }

    @Override
    public Object execute(Map<String, Object> arguments) {
        String command = String.valueOf(arguments.getOrDefault("command", "")).trim();
        if (command.isEmpty()) {
            return Map.of("success", false, "error", "command 不能为空");
        }
        if (!isAllowed(command)) {
            return Map.of("success", false,
                    "error", "命令 '" + firstWord(command) + "' 不在允许列表中，仅支持只读/信息类命令");
        }
        if (SENSITIVE.matcher(command).find()) {
            return Map.of("success", false, "error", "命令包含敏感文件路径（.env / 密钥文件等），已拒绝执行");
        }
        int timeout = Math.min(Math.max(parseInt(arguments.get("timeout"), 10), 1), 30);
        try {
            Process process = new ProcessBuilder(command.split("\\s+")).start();
            if (!process.waitFor(timeout, TimeUnit.SECONDS)) {
                process.destroyForcibly();
                return Map.of("success", false, "error", "命令超时（>" + timeout + "s）");
            }
            String output = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            String error = new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            String body = (output != null && !output.isBlank() ? output : error).trim();
            if (body.isEmpty()) {
                body = "(无输出)";
            }
            if (body.length() > 2000) {
                body = body.substring(0, 2000) + "\n...(输出已截断)";
            }
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("success", process.exitValue() == 0);
            result.put("returncode", process.exitValue());
            result.put("output", body);
            return result;
        } catch (Exception e) {
            return Map.of("success", false, "error", e.getMessage());
        }
    }

    private static boolean isAllowed(String command) {
        String c = command.toLowerCase();
        for (String prefix : ALLOWED_PREFIXES) {
            if (c.startsWith(prefix)) {
                return true;
            }
        }
        return false;
    }

    private static String firstWord(String command) {
        return command.split("\\s+")[0];
    }

    private static int parseInt(Object value, int defaultValue) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        try {
            return Integer.parseInt(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }
}
